xmppserver/src/main/java/org/jivesoftware/openfire/archive/Archiver.java

```diff
@@ -0,0 +1,267 @@
+        while ( running )
+        {
+            // The batch of work for this iteration.
+            new ArrayList<>();
+
+            try
+            {
+                // Blocks until work is produced.
+                ArchiveCandidate<E> work;
+
+                // Continue filling up this batch as long as new archive candidates can be retrieved pretty much
+                // instantaneously, but don't take longer than the maximum allowed purge interval (this is intended
+                // to make sure that the database content is updated regularly)
+                final Instant lastLoopStart = Instant.now();
+                final Duration maxLoopDuration = maxPurgeInterval.minus( gracePeriod );
+                Duration runtime = Duration.ZERO;
+                while ( ( workQueue.size() < maxWorkQueueSize ) // Don't allow the batch to grow to big.
+                    && ( runtime.compareTo( maxLoopDuration ) < 0 ) // Don't take to long between commits.
+                    && ( ( work = queue.poll( gracePeriod.toMillis(), TimeUnit.MILLISECONDS ) ) != null ) )
+                {
+                    workQueue.add( work );
+                    runtime = Duration.between( lastLoopStart, Instant.now() );
+                }
+            }
+            catch ( InterruptedException e )
+            {
+                // Causes the thread to stop.
+                running = false;
+            }
+
+            if ( !workQueue.isEmpty() )
+            {
+                final List<E> batch = workQueue.stream()
+                    .map( ArchiveCandidate::getElement )
+                    .collect( Collectors.toList() );
+                store( batch );
+                lastProcessed = workQueue.get( workQueue.size() -1 ).createdAt();
+                Log.trace( "Stored all produced work in the database. Work size: {}", workQueue.size() );
+                workQueue.clear();
+            }
+        }
+    }
+
+    public void stop()
+    {
+        running = false;
+    }
+
+    /**
+     * Returns an estimation on how long it takes for all data that arrived before a certain instant will have become
+     * available in the data store. When data is immediately available, 'zero', is returned;
+     *
+     * This method is intended to be used to determine if it's safe to construct an answer (based on database
+     * content) to a request for archived data. Such response should only be generated after all data that was
+     * queued before the request arrived has been written to the database.
+     *
+     * This method performs a cluster-wide check, unlike {@link #availabilityETAOnLocalNode(Instant)}.
+     *
+     * @param instant The timestamp of the data that should be available (cannot be null).
+     * @return A period of time, zero when the requested data is already available.
+     */
+    public Duration availabilityETA( Instant instant )
```

Comment:
This is possibly my mis-understanding of what's going on, but this method appears to return either 
a) The ETA on this node, if not zero, or
b) The max ETA on all the other nodes

Should it simply be the max ETA of all nodes? In a two-node cluster, if this node's ETA is 10 seconds, the other nodes ETA is 20 seconds, the overall ETA is 20 seconds, not 10? Should simply be a case of appending localNodeETA to the collection of other node ETAs

Reply:
Your description is correct. I did this on purpose: when it can be determined that the ETA is non-zero, the archive request can't be answered in any case. Adding the overhead of a cluster call, only to get the ETA potentially better, seems like overkill.

Reply:
Hmm, depending on the usage of this method would it better be a boolean, isAllDataArchived() ? Currently a non-zero Duration could well be wrong, which will no doubt be misleading one day in the future, if not right now.

Reply:
That's what the initial signature of this method was. 😁 My concern was that that needed an arbitrary timeout for subsequent checks, which could lead to excessive amounts of checks, depending configuration and load. The ETA intends to replace that with a somewhat OKish estimation. Would putting more emphasis on the "this is an estimate, at best" nature of the return value put your mind at ease?

Reply:
😁 
Yes, I think that should cover it 😄

Reply:
I've now added such a comment.